File: services/api-gateway/app/schemas/hr.py
import uuid
from datetime import date, datetime, time

from pydantic import BaseModel, Field

# ── Staff Profile ───────────────────────────────────────────────────────────


class StaffProfileCreate(BaseModel):
    """Schema for creating/updating a staff profile."""

    staff_id: uuid.UUID
    date_of_birth: date | None = None
    gender: str | None = Field(None, pattern=r"^(male|female|other)$")
    national_id: str | None = Field(None, max_length=20)
    kra_pin: str | None = Field(None, max_length=20)
    nhif_number: str | None = Field(None, max_length=20)
    nssf_number: str | None = Field(None, max_length=20)
    address: str | None = Field(None, max_length=500)
    county: str | None = Field(None, max_length=50)
    sub_county: str | None = Field(None, max_length=50)
    emergency_contact_name: str | None = Field(None, max_length=200)
    emergency_contact_phone: str | None = Field(None, max_length=20)
    emergency_contact_relationship: str | None = Field(None, max_length=50)
    employment_type: str = Field(
        default="permanent",
        pattern=r"^(permanent|contract|casual|intern|locum)$",
    )
    employment_date: date | None = None
    contract_end_date: date | None = None
    probation_end_date: date | None = None
    basic_salary: int | None = Field(None, ge=0)
    allowances: dict | None = None
    qualifications: list[dict] | None = None
    certifications: list[dict] | None = None
    notes: str | None = Field(None, max_length=2000)


class StaffProfileResponse(BaseModel):
    """Schema for staff profile API responses."""

    id: uuid.UUID
    staff_id: uuid.UUID
    date_of_birth: date | None
    gender: str | None
    national_id: str | None
    kra_pin: str | None
    nhif_number: str | None
    nssf_number: str | None
    address: str | None
    county: str | None
    sub_county: str | None
    emergency_contact_name: str | None
    emergency_contact_phone: str | None
    emergency_contact_relationship: str | None
    employment_type: str
    employment_date: date | None
    contract_end_date: date | None
    probation_end_date: date | None
    basic_salary: int | None
    allowances: dict | None
    qualifications: list[dict] | None
    certifications: list[dict] | None
    annual_leave_balance: int
    sick_leave_balance: int
    maternity_leave_balance: int
    paternity_leave_balance: int
    photo_url: str | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class StaffDirectoryItem(BaseModel):
    """Staff member in directory listing with key details."""

    id: uuid.UUID
    employee_number: str
    first_name: str
    last_name: str
    title: str | None
    role: str
    specialization: str | None
    department_name: str | None = None
    phone: str | None
    email: str
    is_active: bool
    license_number: str | None

    model_config = {"from_attributes": True}


class StaffDirectoryResponse(BaseModel):
    """Paginated staff directory response."""

    items: list[StaffDirectoryItem]
    total: int


# ── Shifts ──────────────────────────────────────────────────────────────────


class ShiftCreate(BaseModel):
    """Schema for creating a shift definition."""

    code: str = Field(..., max_length=20)
    name: str = Field(..., max_length=100)
    start_time: time
    end_time: time
    duration_hours: int = Field(..., ge=1, le=24)
    is_night_shift: bool = False
    department_id: uuid.UUID | None = None
    notes: str | None = Field(None, max_length=500)


class ShiftResponse(BaseModel):
    """Schema for shift API responses."""

    id: uuid.UUID
    code: str
    name: str
    start_time: time
    end_time: time
    duration_hours: int
    is_night_shift: bool
    department_id: uuid.UUID | None
    is_active: bool
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ShiftAssignmentCreate(BaseModel):
    """Schema for assigning a staff member to a shift."""

    staff_id: uuid.UUID
    shift_id: uuid.UUID
    department_id: uuid.UUID | None = None
    assignment_date: date
    notes: str | None = Field(None, max_length=500)


class ShiftAssignmentResponse(BaseModel):
    """Schema for shift assignment API responses."""

    id: uuid.UUID
    staff_id: uuid.UUID
    staff_name: str | None = None
    shift_id: uuid.UUID
    shift_name: str | None = None
    department_id: uuid.UUID | None
    assignment_date: date
    status: str
    swapped_with_id: uuid.UUID | None
    swap_reason: str | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ShiftAssignmentListResponse(BaseModel):
    """Paginated shift assignment list."""

    items: list[ShiftAssignmentResponse]
    total: int


# ── Leave ───────────────────────────────────────────────────────────────────


class LeaveRequestCreate(BaseModel):
    """Schema for creating a leave request."""

    leave_type: str = Field(
        ...,
        pattern=r"^(annual|sick|maternity|paternity|compassionate|study|unpaid)$",
    )
    start_date: date
    end_date: date
    days_requested: int = Field(..., ge=1, le=365)
    reason: str | None = Field(None, max_length=1000)
    handover_to: uuid.UUID | None = None
    handover_notes: str | None = Field(None, max_length=1000)


class LeaveApprovalRequest(BaseModel):
    """Schema for approving or rejecting a leave request."""

    action: str = Field(..., pattern=r"^(approve|reject)$")
    rejection_reason: str | None = Field(None, max_length=500)


class LeaveRequestResponse(BaseModel):
    """Schema for leave request API responses."""

    id: uuid.UUID
    staff_id: uuid.UUID
    staff_name: str | None = None
    leave_type: str
    start_date: date
    end_date: date
    days_requested: int
    reason: str | None
    status: str
    approved_by: uuid.UUID | None
    approved_at: datetime | None
    rejection_reason: str | None
    handover_to: uuid.UUID | None
    handover_notes: str | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class LeaveRequestListResponse(BaseModel):
    """Paginated leave request list."""

    items: list[LeaveRequestResponse]
    total: int


# ── Attendance ──────────────────────────────────────────────────────────────


class AttendanceClockIn(BaseModel):
    """Schema for clocking in."""

    shift_id: uuid.UUID | None = None
    notes: str | None = Field(None, max_length=500)


class AttendanceClockOut(BaseModel):
    """Schema for clocking out."""

    notes: str | None = Field(None, max_length=500)


class AttendanceResponse(BaseModel):
    """Schema for attendance API responses."""

    id: uuid.UUID
    staff_id: uuid.UUID
    staff_name: str | None = None
    shift_id: uuid.UUID | None
    shift_name: str | None = None
    attendance_date: date
    clock_in: datetime | None
    clock_out: datetime | None
    status: str
    overtime_minutes: int
    notes: str | None

    model_config = {"from_attributes": True}


class AttendanceListResponse(BaseModel):
    """Paginated attendance list."""

    items: list[AttendanceResponse]
    total: int


# ── Summary ─────────────────────────────────────────────────────────────────


class HRSummary(BaseModel):
    """HR module dashboard summary."""

    total_staff: int = 0
    active_staff: int = 0
    doctors: int = 0
    nurses: int = 0
    on_duty_today: int = 0
    on_leave_today: int = 0
    pending_leave_requests: int = 0
    expiring_contracts: int = 0

File: packages/shared/src/types/hr.ts
/** Staff role */
export type StaffRole =
  | "system_admin"
  | "hospital_admin"
  | "doctor"
  | "nurse"
  | "registration"
  | "records"
  | "cashier"
  | "billing_clerk"
  | "lab_personnel"
  | "pathologist"
  | "pharmacist"
  | "radiologist"
  | "store_keeper"
  | "midwife"
  | "hr_manager"
  | "claims_officer";

/** Employment type */
export type EmploymentType =
  | "permanent"
  | "contract"
  | "casual"
  | "intern"
  | "locum";

/** Leave type */
export type LeaveType =
  | "annual"
  | "sick"
  | "maternity"
  | "paternity"
  | "compassionate"
  | "study"
  | "unpaid";

/** Leave request status */
export type LeaveStatus = "pending" | "approved" | "rejected" | "cancelled";

/** Attendance status */
export type AttendanceStatus =
  | "present"
  | "absent"
  | "late"
  | "half_day"
  | "on_leave"
  | "holiday";

/** Shift assignment status */
export type ShiftAssignmentStatus =
  | "assigned"
  | "confirmed"
  | "completed"
  | "swapped"
  | "cancelled";

/** Staff directory item */
export interface StaffDirectoryItem {
  id: string;
  employee_number: string;
  first_name: string;
  last_name: string;
  title: string | null;
  role: StaffRole;
  specialization: string | null;
  department_name: string | null;
  phone: string | null;
  email: string;
  is_active: boolean;
  license_number: string | null;
}

/** Staff directory response */
export interface StaffDirectoryResponse {
  items: StaffDirectoryItem[];
  total: number;
}

/** Staff profile response */
export interface StaffProfileResponse {
  id: string;
  staff_id: string;
  date_of_birth: string | null;
  gender: string | null;
  national_id: string | null;
  kra_pin: string | null;
  nhif_number: string | null;
  nssf_number: string | null;
  address: string | null;
  county: string | null;
  sub_county: string | null;
  emergency_contact_name: string | null;
  emergency_contact_phone: string | null;
  emergency_contact_relationship: string | null;
  employment_type: EmploymentType;
  employment_date: string | null;
  contract_end_date: string | null;
  probation_end_date: string | null;
  basic_salary: number | null;
  allowances: Record<string, number> | null;
  qualifications: Array<Record<string, string>> | null;
  certifications: Array<Record<string, string>> | null;
  annual_leave_balance: number;
  sick_leave_balance: number;
  maternity_leave_balance: number;
  paternity_leave_balance: number;
  photo_url: string | null;
  notes: string | null;
  created_at: string;
}

/** Staff profile create/update */
export interface StaffProfileCreate {
  staff_id: string;
  date_of_birth?: string | null;
  gender?: string | null;
  national_id?: string | null;
  kra_pin?: string | null;
  nhif_number?: string | null;
  nssf_number?: string | null;
  address?: string | null;
  county?: string | null;
  sub_county?: string | null;
  emergency_contact_name?: string | null;
  emergency_contact_phone?: string | null;
  emergency_contact_relationship?: string | null;
  employment_type?: EmploymentType;
  employment_date?: string | null;
  contract_end_date?: string | null;
  basic_salary?: number | null;
  allowances?: Record<string, number> | null;
  qualifications?: Array<Record<string, string>> | null;
  certifications?: Array<Record<string, string>> | null;
  notes?: string | null;
}

/** Shift definition */
export interface ShiftResponse {
  id: string;
  code: string;
  name: string;
  start_time: string;
  end_time: string;
  duration_hours: number;
  is_night_shift: boolean;
  department_id: string | null;
  is_active: boolean;
  notes: string | null;
  created_at: string;
}

/** Shift create */
export interface ShiftCreate {
  code: string;
  name: string;
  start_time: string;
  end_time: string;
  duration_hours: number;
  is_night_shift?: boolean;
  department_id?: string | null;
  notes?: string | null;
}

/** Shift assignment response */
export interface ShiftAssignmentResponse {
  id: string;
  staff_id: string;
  staff_name: string | null;
  shift_id: string;
  shift_name: string | null;
  department_id: string | null;
  assignment_date: string;
  status: ShiftAssignmentStatus;
  swapped_with_id: string | null;
  swap_reason: string | null;
  notes: string | null;
  created_at: string;
}

/** Shift assignment create */
export interface ShiftAssignmentCreate {
  staff_id: string;
  shift_id: string;
  department_id?: string | null;
  assignment_date: string;
  notes?: string | null;
}

/** Shift assignment list response */
export interface ShiftAssignmentListResponse {
  items: ShiftAssignmentResponse[];
  total: number;
}

/** Leave request response */
export interface LeaveRequestResponse {
  id: string;
  staff_id: string;
  staff_name: string | null;
  leave_type: LeaveType;
  start_date: string;
  end_date: string;
  days_requested: number;
  reason: string | null;
  status: LeaveStatus;
  approved_by: string | null;
  approved_at: string | null;
  rejection_reason: string | null;
  handover_to: string | null;
  handover_notes: string | null;
  notes: string | null;
  created_at: string;
}

/** Leave request create */
export interface LeaveRequestCreate {
  leave_type: LeaveType;
  start_date: string;
  end_date: string;
  days_requested: number;
  reason?: string | null;
  handover_to?: string | null;
  handover_notes?: string | null;
}

/** Leave approval request */
export interface LeaveApprovalRequest {
  action: "approve" | "reject";
  rejection_reason?: string | null;
}

/** Leave request list response */
export interface LeaveRequestListResponse {
  items: LeaveRequestResponse[];
  total: number;
}

/** Attendance response */
export interface AttendanceResponse {
  id: string;
  staff_id: string;
  staff_name: string | null;
  shift_id: string | null;
  shift_name: string | null;
  attendance_date: string;
  clock_in: string | null;
  clock_out: string | null;
  status: AttendanceStatus;
  overtime_minutes: number;
  notes: string | null;
}

/** Attendance clock in */
export interface AttendanceClockIn {
  shift_id?: string | null;
  notes?: string | null;
}

/** Attendance clock out */
export interface AttendanceClockOut {
  notes?: string | null;
}

/** Attendance list response */
export interface AttendanceListResponse {
  items: AttendanceResponse[];
  total: number;
}

/** HR module summary */
export interface HRSummary {
  total_staff: number;
  active_staff: number;
  doctors: number;
  nurses: number;
  on_duty_today: number;
  on_leave_today: number;
  pending_leave_requests: number;
  expiring_contracts: number;
}
